hamilton/experimental/h_cache.py

```diff
@@ -0,0 +1,136 @@
+        **kwargs,
+    ):
+        """Constructs the adapter.
+
+        Parameters
+        ----------
+        cache_path : str
+            Path to the directory where cached files are stored.
+        force_compute : optional, set
+            Set of nodes that should be forced to compute even if cache exists.
+        writers : optional, dict
+            A dictionary of writers for custom formats.
+        readers : optional, dict
+            A dictionary of readers for custom formats.
+        """
+
+        super().__init__(*args, **kwargs)
+        self.cache_path = cache_path
+        self.force_compute = force_compute if force_compute is not None else {}
+        self.computed_nodes = set()
+
+        self.writers = writers or {}
+        self.readers = readers or {}
+
+        self._init_default_readers_writers()
+
+    def _init_default_readers_writers(self):
+        if "json" not in self.writers:
+            self.writers["json"] = write_json
+        if "json" not in self.readers:
+            self.readers["json"] = read_json
+
+        if "feather" not in self.writers:
+            self.writers["feather"] = write_feather
+        if "feather" not in self.readers:
+            self.readers["feather"] = read_feather
+
+        if "parquet" not in self.writers:
+            self.writers["parquet"] = write_parquet
+        if "parquet" not in self.readers:
+            self.readers["parquet"] = read_parquet
+
+    def _check_format(self, fmt):
+        if fmt not in self.writers:
+            raise ValueError(f"invalid cache format: {fmt}")
+
+    def _write_cache(self, data: Any, fmt: str, filepath: str) -> None:
+        self._check_format(fmt)
+        self.writers[fmt](data, filepath)
+
+    def _read_cache(self, fmt: str, filepath: str) -> None:
+        self._check_format(fmt)
+        return self.readers[fmt](filepath)
+
+    def execute_node(self, node: Node, kwargs: dict[str, Any]) -> Any:
+        cache_format = node.tags.get("cache")
+        implicitly_forced = any(dep.name in self.computed_nodes for dep in node.dependencies)
+        if cache_format is not None:
+            filepath = f"{self.cache_path}/{node.name}.{cache_format}"
+            explicitly_forced = node.name in self.force_compute
+            if explicitly_forced or implicitly_forced or not os.path.exists(filepath):
+                result = node.callable(**kwargs)
+                self._write_cache(result, cache_format, filepath)
+                self.computed_nodes.add(node.name)
+                return result
+            return self._read_cache(cache_format, filepath)
+
+        if implicitly_forced:
+            # For purposes of caching, we only mark it as computed if any cached input was computed.
+            # Otherwise, dependants would always be recomputed if they have a non-cached dependency.
+            self.computed_nodes.add(node.name)
+        return node.callable(**kwargs)
```

Comment:
Just to make sure I'm understanding this, this will:

1. Skip computation of a node, if it's cached, independent of input values.
2. It will compute a node with a cache annotation if:
    - the cache file does not exist
    - it was explicitly told to compute it via explicitly_forced
    - if there was an earlier cached node that was upstream that was recomputed via implicitly_forced

This means that a user operating this would:
 - stick @cache on say data loading steps -- usually the beginning of the DAG
 - stick @cache on anything else
 - then as they iterate, either force re-computation, or blow away the directory to run and cache new values

Right?

Reply:
Yeah, this sounds right. One case that is missing (I'll need to double-check, but at the first glance it seems true) is that a node will be computed if an upstream node was computed due to missing file, not only if it was forced.

I'm not entirely sure if this is desirable, though. It's definitely erring on the side of caution (it _is_ computed, so there's no guarantee that the downstream shouldn't change, e.g., because it's reading an updated dataset), but not quite sure if this would be an intuitive behavior for everyone.

Any thoughts on that? In the usecase I was writing it for, the current way is the expected one, but doesn't mean it would be always.

We could always give an option to do one way or another but this also further complicates the interface and this could be difficult to express the intention (it took me a paragraph here to explain).

Reply:
Ah yes. I would just mention that as part of the caveats; I wouldn't build for that case.
To that end, we could add `debug` logs that people can turn on to see what was or was not loaded from cache.